File: app/src/main/java/com/example/waveplayer/fragments/fragment_playlists/RecyclerViewAdapterPlaylists.java
package com.example.waveplayer.fragments.fragment_playlists;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.RecyclerView;

import com.example.waveplayer.activity_main.ActivityMain;
import com.example.waveplayer.R;
import com.example.waveplayer.random_playlist.RandomPlaylist;

import java.util.List;

public class RecyclerViewAdapterPlaylists extends RecyclerView.Adapter<RecyclerViewAdapterPlaylists.ViewHolder> {

    private final FragmentPlaylists fragmentPlaylists;

    public List<RandomPlaylist> randomPlaylists;

    private OnCreateContextMenuListenerPlaylists onCreateContextMenuListenerPlaylists;

    private View.OnClickListener onClickListenerHandle;

    private View.OnClickListener onClickListenerViewHolder;

    public RecyclerViewAdapterPlaylists(FragmentPlaylists fragmentPlaylists, List<RandomPlaylist> items) {
        this.fragmentPlaylists = fragmentPlaylists;
        randomPlaylists = items;
    }

    public void updateList(List<RandomPlaylist> randomPlaylists) {
        this.randomPlaylists = randomPlaylists;
        notifyDataSetChanged();
    }

    @Override
    @NonNull
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_playlist, parent, false));
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        holder.randomPlaylist = randomPlaylists.get(position);
        holder.textViewPlaylistName.setText(randomPlaylists.get(position).getName());
        holder.handle.setOnCreateContextMenuListener(null);
        onCreateContextMenuListenerPlaylists =
                new OnCreateContextMenuListenerPlaylists(fragmentPlaylists, holder.randomPlaylist);
        holder.handle.setOnCreateContextMenuListener(onCreateContextMenuListenerPlaylists);
        onClickListenerHandle = v -> holder.handle.performLongClick();
        holder.handle.setOnClickListener(null);
        holder.handle.setOnClickListener(onClickListenerHandle);
        onClickListenerViewHolder = v -> {
            int position1 = holder.getAdapterPosition();
            if (position1 != RecyclerView.NO_POSITION) {
                ActivityMain activityMain = ((ActivityMain) fragmentPlaylists.getActivity());
                if (activityMain != null) {
                    fragmentPlaylists.setUserPickedPlaylist(holder.randomPlaylist);
                }
                NavHostFragment.findNavController(fragmentPlaylists).navigate(
                        FragmentPlaylistsDirections.actionFragmentPlaylistsToFragmentPlaylist());
            }
        };
        holder.playlistView.setOnClickListener(null);
        holder.playlistView.setOnClickListener(onClickListenerViewHolder);
    }

    @Override
    public void onViewRecycled(@NonNull ViewHolder holder) {
        super.onViewRecycled(holder);
        onCreateContextMenuListenerPlaylists = null;
        holder.handle.setOnCreateContextMenuListener(null);
        onClickListenerHandle = null;
        holder.handle.setOnClickListener(null);
        onClickListenerViewHolder = null;
        holder.playlistView.setOnClickListener(null);
        holder.randomPlaylist = null;
    }

    @Override
    public int getItemCount() {
        return randomPlaylists.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        final public View playlistView;
        final public TextView textViewPlaylistName;
        public RandomPlaylist randomPlaylist;
        final ImageView handle;

        public ViewHolder(View view) {
            super(view);
            playlistView = view;
            textViewPlaylistName = view.findViewById(R.id.text_view_playlist_name);
            if (randomPlaylist != null) {
                textViewPlaylistName.setText(randomPlaylist.getName());
            }
            handle = view.findViewById(R.id.playlist_handle);
        }

        @Override
        @NonNull
        public String toString() {
            return randomPlaylist.getName();
        }

    }

}

File: app/src/main/java/io/fourth_finger/pinky_player/fragments/RecyclerViewAdapterPlaylists.kt
package io.fourth_finger.pinky_player.fragments

import android.annotation.SuppressLint
import android.view.ContextMenu
import android.view.ContextMenu.ContextMenuInfo
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import io.fourth_finger.pinky_player.R
import io.fourth_finger.playlist_data_source.RandomPlaylist

class RecyclerViewAdapterPlaylists(
    private var listenerCallbackPlaylists: ListenerCallbackPlaylists,
    private var randomPlaylists: List<RandomPlaylist>
) : RecyclerView.Adapter<RecyclerViewAdapterPlaylists.ViewHolder>() {


    interface ListenerCallbackPlaylists {
        fun onClickViewHolder(randomPlaylist: RandomPlaylist)
        fun onMenuItemClickAddToPlaylist(randomPlaylist: RandomPlaylist)
        fun onMenuItemClickAddToQueue(randomPlaylist: RandomPlaylist)
    }

    @SuppressLint("NotifyDataSetChanged")
    fun updateList(randomPlaylists: List<RandomPlaylist>) {
        this.randomPlaylists = randomPlaylists
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(
            LayoutInflater.from(parent.context).inflate(
                R.layout.item_playlist,
                parent,
                false
            )
        )
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.randomPlaylist = randomPlaylists[position]
        holder.textViewPlaylistName.text = randomPlaylists[position].name
        holder.handle.setOnCreateContextMenuListener { menu: ContextMenu?, _: View?, _: ContextMenuInfo? ->
            val itemAddToPlaylist = menu?.add(R.string.add_to_playlist)
            itemAddToPlaylist?.setOnMenuItemClickListener {
                listenerCallbackPlaylists.onMenuItemClickAddToPlaylist(randomPlaylists[position])
                true
            }
            val itemAddToQueue = menu?.add(R.string.add_to_queue)
            itemAddToQueue?.setOnMenuItemClickListener {
                listenerCallbackPlaylists.onMenuItemClickAddToQueue(randomPlaylists[position])
                true
            }
        }
        holder.handle.setOnClickListener { holder.handle.performLongClick() }
        holder.playlistView.setOnClickListener {
                listenerCallbackPlaylists.onClickViewHolder(randomPlaylists[position])
        }
    }

    override fun getItemCount(): Int {
        return randomPlaylists.size
    }

    class ViewHolder(val playlistView: View) : RecyclerView.ViewHolder(playlistView) {
        val textViewPlaylistName: TextView = playlistView.findViewById(R.id.text_view_playlist_name)
        val handle: ImageView = playlistView.findViewById(R.id.playlist_handle)

        lateinit var randomPlaylist: RandomPlaylist

        override fun toString(): String {
            return randomPlaylist.name
        }

    }

}